pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionAssignmentTailer.java

```diff
@@ -18,56 +18,91 @@
  */
 package org.apache.pulsar.functions.worker;
 
-import java.io.IOException;
-import java.util.function.Function;
-
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
 import org.apache.pulsar.client.api.Message;
-import org.apache.pulsar.client.api.PulsarClientException.AlreadyClosedException;
+import org.apache.pulsar.client.api.MessageId;
+import org.apache.pulsar.client.api.PulsarClientException;
 import org.apache.pulsar.client.api.Reader;
-import org.apache.pulsar.common.util.FutureUtil;
+import org.apache.pulsar.client.api.ReaderBuilder;
 import org.apache.pulsar.functions.proto.Function.Assignment;
 
-import lombok.extern.slf4j.Slf4j;
+import java.io.IOException;
 
 @Slf4j
-public class FunctionAssignmentTailer
-    implements java.util.function.Consumer<Message<byte[]>>, Function<Throwable, Void>, AutoCloseable {
+public class FunctionAssignmentTailer implements AutoCloseable {
 
     private final FunctionRuntimeManager functionRuntimeManager;
+    @Getter
     private final Reader<byte[]> reader;
-    private boolean closed = false;
+    private volatile boolean isRunning = false;
```

Comment:
instead of isRunning, better to word it as stopRunning

Reply:
why do we need to change it?  We would also need to reverse the boolean logic checks to be semantically correct?

Reply:
Yes. isRunning implies that its a state. stopRunning implies its a command.
What we really have here is that the thread should keep running until its commanded by the main thread to stop. Hence the stopRunning instead of isRunning

Reply:
But we also have a method called "start"?

Reply:
At start() {
shouldStop = false;
create the thread and start
}
At close() {
set shoutStop = true, interrupt thread, join and then set it to null;
close reader and set it to null
}

This way I can do start() /close() multiple times